### PR TITLE
Fix the wording of help msg for bitwise not

### LINKED ISSUE #101793: Slight refinement to wording of help suggestion as a result of `not` being used
Given the following code: [link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=6a7c3c414c4a2ed446ef138d60dc11b0)

```rust
fn main() {
    let x = not 123;
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
error: unexpected `123` after identifier
 --> src/main.rs:2:17
  |
2 |     let x = not 123;
  |             ----^^^
  |             |
  |             help: use `!` to perform logical negation

error: could not compile `playground` due to previous error
```

The above phrasing AFAIK should say "bitwise not" instead of "logical negation" as the above literal is not interpreted as a Boolean.
The following shows a slightly different input and output which has the right phrasing: [link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=09fa101b89d0b730f6f5d4a9a17812f3)
```rust
fn main() {
    let x = ~ 123;
}
```

```
   Compiling playground v0.0.1 (/playground)
error: `~` cannot be used as a unary operator
 --> src/main.rs:2:13
  |
2 |     let x = ~ 123;
  |             ^ help: use `!` to perform bitwise not

error: could not compile `playground` due to previous error
``` 
Thanks a lot.

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #101793

### CODE REVIEW

**Comment by Rageking8** [comment]
Ehm shouldn't the phrasing "bitwise not" only be used for integer types and the phrasing "logical negation" be used for expressions that evaluate to a Boolean?

**Comment by chenyukang** [comment]
> Ehm shouldn't the phrasing "bitwise not" only be used for integer types and the phrasing "logical negation" be used for expressions that evaluate to a Boolean?

You are right, let me tuning it more.

**Comment by chenyukang** [comment]
For `not expr ...`,
at the point of parsing, we can not know what the real type of `expr` untile type checking.

so, maybe the proper message is:
`use `!` to perform logical negation or bitwise not` ?

We can try to testing whether next token is a number, but in other scenario, we can not get it accurate for boolean and number.

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -660,9 +660,18 @@ impl<'a> Parser<'a> {
     fn recover_not_expr(&mut self, lo: Span) -> PResult<'a, (Span, ExprKind)> {
         // Emit the error...
         let negated_token = self.look_ahead(1, |t| t.clone());
+        let negtated_msg = if negated_token.is_numeric_lit() {
```
Comment (compiler-errors) [inline compiler/rustc_parse/src/parser/expr.rs:663]: I don't think we should be passing this in as an english phrase -- either use a [fluent selector](https://projectfluent.org/fluent/guide/selectors.html) or split this into several different suggestion messages.

Comment (chenyukang) [reply compiler/rustc_parse/src/parser/expr.rs]: Fixed, thanks @compiler-errors.

**Review by davidtwco — APPROVED** [review]

**Pull request merged**